Block editor: use the official slot API to replace close button

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

Replace the hacky way of overriding the close button with a new API which was added in WordPress/gutenberg#22323.

* Use the `<MainDashboardButton>` component to override the close button.
* Remove old CSS selector method of overriding button
* Add CSS rules to fix button height. Unfortunately the slot API adds a wrapper div to the button, which breaks it's natural height.
* Install `@wordpress/interface` as a dependency so it can be bundled. This package doesn't appear to be available on the `wp` global, and there doesn't appear to be an enqueue script parameter that will add it.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout branch, run dev build of calypso, and set up your sandbox for wpcom-block-editor: PCYsg-l4k-p2
* Clicking button should navigate back to Calypso
* If the document is dirty you should be warned and allowed to cancel the navigation
* The tooltip label should actually says the place you're going back to (this wasn't possible before 🎉)
* Button takes you back to somewhere other than customer home:
  * Activate a new theme
  * In dialog that appears click "edit my homepage"
  * Block editor opens correctly
  * Close editor, you should be taken back to the theme gallery
* Test atomic site, with and without gutenberg plugin activated
* Test that legacy FSE site isn't affected (it should still have a back button with a chevron)
* Test Jetpack site can navigate back to Calypso

p-jackson committed:
Override close button using <MainDashboardButton> slot

simison commented:
Will this work for Atomic sites running older Gutenberg version or Gutenberg plugin disabled?

vindl commented:
> Will this work for Atomic sites running older Gutenberg version or Gutenberg plugin disabled?

No, since only the newer plugin versions will contain the required slot.

p-jackson commented:
> Will this work for Atomic sites running older Gutenberg version or Gutenberg plugin disabled?

Thanks @simison I hadn't thought of that. The problem of atomic sites with old GB versions isn't unique to this new technique though. The selector used to override the close button also changed sometimes, so it was already happening.

For GB versions < 8.2 the close button just won't get overridden so will take the user back to wp-admin. That seems acceptable to me, what do you think?

p-jackson commented:
When testing this on an atomic site that has GB deactivated, the close button is still redirecting back to calypso 🤔 I guess that's nice, but I'm not sure where the code is that's doing this.

lsl commented:
Change itself LGTM.

> For GB versions < 8.2 the close button just won't get overridden so will take the user back to wp-admin. That seems acceptable to me, what do you think?

+1

### Testing 
- [x] Checkout branch and set up your sandbox: PCYsg-l4k-p2
- [x] Clicking button should navigate back to Calypso
- [x] If the document is dirty you should be warned and allowed to cancel the navigation
- [ ] The tooltip label should actually says the place you're going back to (this wasn't possible before tada)
- [x] Button takes you back to somewhere other than customer home:
- [x] Activate a new theme
- [x] ^ In dialog that appears click "edit my homepage"
- [x] ^ Block editor opens correctly
- [ ] ^ Close editor, you should be taken back to the theme gallery
- [x] Test atomic site, with and without gutenberg plugin activated
- [ ] Test that legacy FSE site isn't affected (it should still have a back button with a chevron)
- [ ] Test Jetpack site can navigate back to Calypso

I'll try and get more of this going later. I couldn't get it to take me back to the theme gallery. Possibly just some caching issues I need to figure out on my end. It didn't have the title labels in the themes case, but did see them in most other cases. When the label is missing it takes me back to the post/page screen. Thinking this might be the current/default behavior?

simison commented:
> > Will this work for Atomic sites running older Gutenberg version or Gutenberg plugin disabled?
> 
> No, since only the newer plugin versions will contain the required slot.

Right, I understand that. :-) I was mainly concerned that it won't fatal and if there's need for some existing overrides to keep functioning.

> When testing this on an atomic site that has GB deactivated, the close button is still redirecting back to calypso 🤔 I guess that's nice, but I'm not sure where the code is that's doing this.

Could be in https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor or in `wpcomsh`.

p-jackson commented:
> It didn't have the title labels in the themes case, but did see them in most other cases.

The label's missing for me sometimes too. It happens when the `calypsoifyGutenberg.closeButtonLabel` global is `undefined`. Not sure why that's happening. I don't think it's a bug introduced by this PR, but I think it's been exposed by this PR, so I'll see why the label isn't always being set.

ramonjd commented:
I tested by running build-wpcom-block-editor build script on my sandbox with the wpcom-block-editor build number 777609

Sandboxing `widgets.wp.com` and running Calypso locally

**Clicking button should navigate back to Calypso** ✅ 

**If the document is dirty you should be warned and allowed to cancel the navigation** ✅ 

**The tooltip label should actually** says the place you're going back to (this wasn't possible before 🎉) ✅  (View Pages|View Posts)

**Test that legacy FSE site isn't affected** (it should still have a back button with a chevron)
<img width="281" alt="Screen Shot 2020-06-18 at 11 33 15 am" src="https://user-images.githubusercontent.com/6458278/84968133-23db4600-b159-11ea-90a9-b013fbf84994.png"> ✅

**Button takes you back to somewhere other than customer home:**
This doesn't work for me, I'm always taken back to View Pages. 
Similarly when I click to edit the home page from Customer Home, I'm not returned to Customer home (assuming that's also intended?).

I have a site running on Gutenberg edge (8.3), and I'm seeing this. My sandbox is clean. I'm not sure if it's related or not.

![Jun-18-2020 11-59-03](https://user-images.githubusercontent.com/6458278/84970728-9d296780-b15e-11ea-8ed7-05057fcf7100.gif)

p-jackson committed:
Calypsoify iframe always sends close label to editor

Now that we're overriding the close button completely we can't full back
on the existing label text. Matched previous label text.

p-jackson committed:
Update close label text if updateCloseButtonOverrides action fires

This deals with case where close button renders before message is
recieved.

a8ci18n commented on client/state/selectors/get-editor-close-config.js:
```diff
@@ -57,7 +57,16 @@ export default function getEditorCloseConfig( state, siteId, postType, fseParent
+
+	let label = translate( 'Back' );
+	if ( postType === 'post' ) {
+		label = translate( 'View posts' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **116** times:
`translate( 'View Posts' )` *ES Score: 8*

a8ci18n commented on client/state/selectors/get-editor-close-config.js:
```diff
@@ -57,7 +57,16 @@ export default function getEditorCloseConfig( state, siteId, postType, fseParent
+	if ( postType === 'post' ) {
+		label = translate( 'View posts' );
+	} else if ( postType === 'page' ) {
+		label = translate( 'View pages' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **115** times:
`translate( 'View Pages' )` *ES Score: 9*

p-jackson committed:
Close button uses <a> tag

This matches the previous behaviour before overriding the button. This
allows users to open link in new tab.

p-jackson commented:
- Fixed label issue. Calypso wasn't sending a label when it was happy to use the existing label in Gutenberg. We always need to provide a label now since we're completely overriding the GB button, we can't fall back to it's default props.
- I've also switched the `<button>` to a `<a>` which is what GB uses. This way the user can still do things like right-click to open in a new tab, or use a screen reader to announce all links on the page.
- Added code which should deal with timing issues, like if the button happened to render before the `calypsoifyGutenberg` global was set.
- I think I've discovered the other code in Jetpack that was overriding the close button that was allowing Atomic sites with GB disabled to still work.
 https://github.com/Automattic/jetpack/blob/563f49ee866f0b005cd7b02d3289e4330a2eb6d4/modules/calypsoify/mods-gutenberg.js#L33
 I think it's harmless, but perhaps it should be updated too to use the new method? @simison 

Another wee test would be appreciated @ramonjd @lsl since it's changed a wee bit. Testing will now require `calypso.localhost:3000`

p-jackson commented:
Here's another place that's overriding the link url 256b7-pb

It might be dead code though because the selector being used doesn't even work with the latest version of GB. Is this for a special case I'm not aware of @vindl? (saw your name in the history 😅)

ramonjd commented:
>  Close editor, you should be taken back to the theme gallery

Ran through the scenarios again and all is looking great. Always a label there 👍 
 
The navigation from themes still doesn't work for me. 

![Jun-19-2020 11-42-21](https://user-images.githubusercontent.com/6458278/85087872-07571080-b222-11ea-9352-b61d8c6b7efc.gif)

I see your changes: the `wpcom-block-editor__close-button` is now an `<a />`, and I'm running this branch locally on http://calypso.localhost:3000/ 

Am I missing something?

p-jackson commented:
@ramonjd that's very strange. Could you check what the value of `calypsoifyGutenberg.closeButtonLabel` and `calypsoifyGutenberg.closeUrl` are after you switch from the theme gallery? It could be that it's all by design and that the feature to return to the theme gallery is an a/b test or feature flag or something.

p-jackson commented:
@ramonjd 😑 figured it out

I activate the theme by going to `https://wordpress.com/themes/mysite.com`, clicking the 3 dots and clicking activate. You activate the theme by going to `https://wordpress.com/theme/twentytwenty/mysite.com` and clicking activate.

The code that sets the close url uses the `/^\/themes\/?/` regex to decide whether to return the user to the theme gallery. And your url doesn't match.

So by design.

andrewserong reviewed (approved):
LGTM! This is so nice with the official slot API. Just noting that this will likely need a rebase before deployment / generating a diff, so that recently merged in `wpcom-block-editor` changes are not overridden.

Also, I noticed that with `define( 'WPCOM_BLOCK_EDITOR_SIDEBAR' );` set in my Sandbox, the new sidebar nav work isn't displaying correctly, but I imagine that's either that the backend component of https://github.com/Automattic/wp-calypso/pull/43471 hasn't been deployed yet, or this will be fixed by https://github.com/Automattic/wp-calypso/pull/43472?

![image](https://user-images.githubusercontent.com/14988353/85251166-f285b180-b49b-11ea-835d-2c0c0c57b53f.png)

The error in the console is `TypeError: Cannot read property 'slug' of undefined`.

Since all that work is still in progress, though, and doesn't seem to be directly related to this change, I'm giving it the ✅ , anyway, but let's see if the rebase resolves it.

Tested:

✅ Clicking button should navigate back to Calypso
If the document is dirty you should be warned and allowed to cancel the navigation
✅ The tooltip label should actually says the place you're going back to (this wasn't possible before 🎉)
✅ Button takes you back to somewhere other than customer home (tested against pages / posts / theme selection)
✅ Test atomic site, with and without gutenberg plugin activated
✅ Test that legacy FSE site isn't affected (it should still have a back button with a chevron)
✅ Test Jetpack site can navigate back to Calypso (I run my Jetpack site on base WordPress.org without the Gutenberg plugin installed, so the other fallback script is good for my use-case, too!)

p-jackson commented:
> I noticed that with define( 'WPCOM_BLOCK_EDITOR_SIDEBAR' ); set in my Sandbox, the new sidebar nav work isn't displaying correctly, but I imagine that's either that the backend component of #43471 hasn't been deployed yet, or this will be fixed by #43472?

Yup that's expected. This PR will be deployed straight to users but `WPCOM_BLOCK_EDITOR_SIDEBAR` is currently a dev only thing. Should support it soon 🤞

sirreal commented:
The new dependency seems to be duplicating wordpress/* dependencies. It would be nice to get https://github.com/Automattic/wp-calypso/pull/42835 landed first.

sirreal commented:
> The new dependency seems to be duplicating wordpress/* dependencies. It would be nice to get #42835 landed first.

This wasn't intended to be a blocker, although it does read like a soft blocker 🙂 

Please just consider the tradeoff (difficulty of getting https://github.com/Automattic/wp-calypso/pull/42835 merged now vs. deduplicating the bundles later) and be sure that the packages that are being duplicated here work well when duplicated. Several @wordpress/* packages have behaved poorly when duplicated in the past. For example, I notice `@wordpress/i18n` is duplicated here, it may be problematic because the translation function exports (`__`, `setLocaleData`) operate on a module singleton, but it's also hard to know exactly how those issues may manifest.

p-jackson commented:
Thanks for the clarification @sirreal :) This holds up some other PRs so would be nice to merge, but I can also understand the conundrum. I might see how straightforward fixing the duplication here is. It might just be a tricky problem regardless of where we fix it.

vindl commented:
> Here's another place that's overriding the link url 256b7-pb
> It might be dead code though because the selector being used doesn't even work with the latest version of GB. Is this for a special case I'm not aware of @vindl? (saw your name in the history 😅)

I thought it might be a legacy thing for Jetpack/Atomic, since older Gutenberg versions could run there but that doesn't seem to be the case: https://github.com/Automattic/jetpack/blob/0c9cd3e493d4b0e7ee7ea8d96710dde93e92de85/modules/calypsoify/mods-gutenberg.js

I don't remember what special case that might be, so if it smoke tests fine without it I think it should be fine to remove.

p-jackson commented:
Rebased and re-installed dependencies to make sure it was done cleanly. Double checked the duplicated dependencies and (after the update) all duplicates are already present on the main branch. Will release this since it's not taking us backwards.
